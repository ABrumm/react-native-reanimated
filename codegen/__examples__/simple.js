
const ABS = `
import { cond, lessThan, multiply, codegen } from '../base';

const abs = codegen('abs')((a) => {
  return cond(lessThan(a, 0), multiply(-1, a), a);
})

export default abs;
`.trim();

export default {
  ABS
}
