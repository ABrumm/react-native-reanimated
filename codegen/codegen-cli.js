const codegen = require("./codegen");

const INPUT = [
  "./src/derived/abs.js"
]

const JS_MODULE = "./codegen/jsModule.js"
const IOS = "./ios/REANodesManager.m"


const fs = require('fs');



function readFileAsync(filename, enc) {
  return new Promise(function(resolve, reject) {
    fs.readFile(filename, 'utf8', function(err, data){
      if (err)
        reject(err);
      else
        resolve(data);
    });
  });
};


async function traverse() {
  const ioscontent = await readFileAsync(IOS)
  const inputs = [];
  for (const i of INPUT) {
    inputs.push(codegen(await readFileAsync(i)))
  }
  const jscontent = await readFileAsync(JS_MODULE)
  
  console.log(jscontent)
  
  // TODO
  const { objc, js, registry, objcregistry } = inputs[0];
  
  js.replace("// REGISTRY", registry)
  
  const codegenedJsModule = jscontent
    .replace("// REGISTRY", registry)
    .replace("// NODES", js)
  
  const codegenedObjcModule = ioscontent
    .replace("// CODEGEN CLASSES", objc)
    .replace("// CODEGEN REGISTER", objcregistry)
  
  
  
  fs.writeFile(IOS, codegenedObjcModule, function(){console.log('done objc')});
  fs.writeFile(JS_MODULE, codegenedJsModule, function(){console.log('done js')});
}

traverse()
