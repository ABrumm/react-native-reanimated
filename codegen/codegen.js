/* eslint-disable no-case-declarations */
const { Parser } = require('acorn');

const JS_CLASS = `
class ::__NODE_NAME__:: extends AnimatedNode {
  constructor(input) {
    super(
      {
        type: '::__NODE_NAME__::',
        input: input,
      },
      input
    );
  }
}
`.trim()

const JS_REGISTRY = `
const registry = {
::__ITEMS__::
}
`.trim()

const JS_REGISTRY_ITEM= `
['::__ID__::']: ::__CLASS__::,
`.trim()

const OBJ_REGISTRY_ITEM= `
@"::__CLASS__::": [REA::__CLASS__::Node class],
`.trim()


const OBJC_TEMPLATE = `
@interface REA::__NODE_NAME__::Node : REANode
@end

@implementation REA::__NODE_NAME__::Node {
  NSArray<NSNumber *> *_input;
}

- (instancetype)initWithID:(REANodeID)nodeID config:(NSDictionary<NSString *,id> *)config
{
  if ((self = [super initWithID:nodeID config:config])) {
    _input = config[@"input"];
  }
  return self;
}

- (id)evaluate
{
  return ::__CODE__::;
}

@end


`.trim()

function jsUcfirst(string)
{
  return string.charAt(0).toUpperCase() + string.slice(1);
}


function traverseNode(astNode, args) {
  if (astNode.type === 'Identifier') {
    return ({
      id: args[astNode.name],
      type: 'Identifier',
    })
  }
  

  if (astNode.type === 'Literal') {
    return ({
      value: astNode.value,
      type: 'Literal',
    })
  }
  
  // -1 makes me really sad
  if (astNode.type === 'UnaryExpression' && astNode.operator === '-') {
    return ({
      value: 0 - astNode.argument.value,
      type: 'Literal',
    })
  }
  
  
  return ({
    name: astNode.callee.name,
    type: 'Generic',
    args: astNode.arguments.map(v => traverseNode(v, args))
  })
}

const OBJC_COND = `
  (::__IF__::) ?
    ::__THEN__::
    : @(0)
`.trim();

const OBJC_COND_ELSE = `
  (::__IF__::) ?
    ::__THEN__::
    : ::__ELSE__::
`.trim()

const OBJC_LESS_THEN = `
  ::__A__:: < ::__B__::
`.trim()

const OBJC_MULTIPLY = `
  @(::__A__:: * ::__B__::)
`.trim()

const OBJC_NUMBER_LITERAL = `
  ::__A__::
`.trim()

const OBJC_IDENTIFIER = `
  [[self.nodesManager findNodeByID:_input[::__ID__::]] value]
`.trim()

const OBJC_NUM_IDENTIFIER = `
  [${OBJC_IDENTIFIER} doubleValue]
`.trim()




function codegenNode(node, enforceNumerical = false) {
  if (node.type === 'Literal') {
    // todo: handle string or undef
    return OBJC_NUMBER_LITERAL.replace('::__A__::', node.value)
  }
  
  if (node.type === 'Identifier') {
    if (enforceNumerical) {
      return OBJC_NUM_IDENTIFIER.replace('::__ID__::', node.id)
    }
    return OBJC_IDENTIFIER.replace('::__ID__::', node.id)
  }
  switch (node.name) {
    case 'cond':
      const template = node.args[2] ? OBJC_COND_ELSE : OBJC_COND;
      return template
        .replace('::__IF__::', codegenNode(node.args[0]))
        .replace('::__THEN__::', codegenNode(node.args[1]))
        .replace('::__ELSE__::', codegenNode(node.args[2]));
    case 'lessThan':
      return OBJC_LESS_THEN
        .replace('::__A__::', codegenNode(node.args[0], true))
        .replace('::__B__::', codegenNode(node.args[1], true))
  
    case 'multiply':
      return OBJC_MULTIPLY
        .replace('::__A__::', codegenNode(node.args[0], true))
        .replace('::__B__::', codegenNode(node.args[1], true))
  }
  return null;
}


module.exports = function codegen(input) {
  const ast =  Parser.parse(input, { sourceType: "module" })
  const red =
    ast.body
      .filter(v => v.type === "VariableDeclaration")
      .map(v => v.declarations[0])
      .filter(v => v.init.type === 'CallExpression')
      .filter(v => v.init.callee.callee.name === 'codegen')
      .filter(v => v.init.callee.arguments[0].type === 'Literal')
      .map(v => {
        const args = v.init.arguments[0].params.map(v => v.name).reduce((curr, w, i) => ({ ...curr, [w]: i }), {})
        return ({
          ret: traverseNode(v.init.arguments[0].body.body.find(w => w.type === "ReturnStatement").argument, args),
          name: 'CODEGEN' + jsUcfirst(v.init.callee.arguments[0].value),
          id: v.init.callee.arguments[0].value
        })
      })
  const objc = red
    .map(
      v => OBJC_TEMPLATE
        .replace(/::__NODE_NAME__::/g, v.name)
        .replace(/::__CODE__::/g, codegenNode(v.ret))
    )
    .join('\n')
  
  const js = red
    .map(
      v => JS_CLASS
        .replace(/::__NODE_NAME__::/g, v.name)
    )
    .join('\n')
  
  const registry = JS_REGISTRY.replace('::__ITEMS__::',
    red
      .map(
        v => JS_REGISTRY_ITEM
          .replace(/::__ID__::/g, v.id)
          .replace(/::__CLASS__::/g, v.name)
      )
      .join('\n')
  )
  
  
  const objcregistry = red
    .map(
      v => OBJ_REGISTRY_ITEM
        .replace(/::__CLASS__::/g, v.name)
    )
    .join('\n')
  
  
  return { objc, js, registry, objcregistry }
}

