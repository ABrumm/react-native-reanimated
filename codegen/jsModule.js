// eslint-disable-next-line no-unused-vars
import AnimatedNode from "../src/core/AnimatedNode";

// NODES

// REGISTRY

export default function codegen(name) {
  return function (...args) {
    return function (...args) {
      // eslint-disable-next-line no-undef
      return new registry[name](args.map(a => a.__nodeID))
    }
  }
}
