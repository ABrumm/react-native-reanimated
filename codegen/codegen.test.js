import codegen from './codegen'
import simple from "./__examples__/simple";
describe('Generates objc', () => {
  Object.keys(simple).map(key => {
    it(`${key} match`, () => {
      expect(
        codegen(simple[key]).objc
      ).toMatchSnapshot();
    })
  })
  
})

describe('Generates js', () => {
  Object.keys(simple).map(key => {
    it(`${key} match`, () => {
      expect(
        codegen(simple[key])
      ).toMatchSnapshot();
    })
  })
  
})
