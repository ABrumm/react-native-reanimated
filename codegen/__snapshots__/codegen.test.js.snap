// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generates js ABS match 1`] = `
Object {
  "js": "class CODEGENAbs extends AnimatedNode {
  constructor(input) {
    super(
      {
        type: 'CODEGENAbs',
        input: input,
      },
      input
    );
  }
}",
  "objc": "@interface REACODEGENAbsNode : REANode
@end

@implementation REACODEGENAbsNode {
  NSArray<NSNumber *> *_input;
}

- (instancetype)initWithID:(REANodeID)nodeID config:(NSDictionary<NSString *,id> *)config
{
  if ((self = [super initWithID:nodeID config:config])) {
    _input = config[@\\"input\\"];
  }
  return self;
}

- (id)evaluate
{
  if ([[[self.nodesManager findNodeByID:_input[0]] value] doubleValue] < 0) {
    return @(-1 * [[[self.nodesManager findNodeByID:_input[0]] value] doubleValue]);
  }
  return [[self.nodesManager findNodeByID:_input[0]] value];

}

@end",
  "objcregistry": "@\\"CODEGENAbs\\": [REACODEGENAbsNode class],",
  "registry": "const registry = {
['abs']: CODEGENAbs,
}",
}
`;

exports[`Generates objc ABS match 1`] = `
"@interface REACODEGENAbsNode : REANode
@end

@implementation REACODEGENAbsNode {
  NSArray<NSNumber *> *_input;
}

- (instancetype)initWithID:(REANodeID)nodeID config:(NSDictionary<NSString *,id> *)config
{
  if ((self = [super initWithID:nodeID config:config])) {
    _input = config[@\\"input\\"];
  }
  return self;
}

- (id)evaluate
{
  if ([[[self.nodesManager findNodeByID:_input[0]] value] doubleValue] < 0) {
    return @(-1 * [[[self.nodesManager findNodeByID:_input[0]] value] doubleValue]);
  }
  return [[self.nodesManager findNodeByID:_input[0]] value];

}

@end"
`;
